lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+var errServicesStopFailed = errors.New("services stop failed")
+
+type ephemeralService interface {
+	Run() error
+	Stop() error
+}
+
+type finalizationHandler struct {
+	servicesLock       sync.Mutex
+	finalizationEngine ephemeralService
+	votingRound        ephemeralService
+
+	newServices   func() (engine, voting ephemeralService)
+	timeout       time.Duration
+	initiateRound func() error
+
+	stopCh      chan struct{}
+	handlerDone chan struct{}
+	errorCh     chan error
+}
+
+func newFinalizationHandler(service *Service) *finalizationHandler {
+	// builder is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	builder := func() (engine, voting ephemeralService) {
+		finalizationEngine := newFinalizationEngine(service)
+		votingRound := newHandleVotingRound(service, finalizationEngine.actionCh)
+		return finalizationEngine, votingRound
+	}
+
+	return &finalizationHandler{
+		newServices: builder,
+
+		timeout:       5 * time.Second,
+		initiateRound: service.initiateRound,
+		stopCh:        make(chan struct{}),
+		handlerDone:   make(chan struct{}),
+		errorCh:       make(chan error),
+	}
+}
+
+func (fh *finalizationHandler) Start() (errorCh <-chan error, err error) {
+	go fh.start()
+	return fh.errorCh, nil
+}
+
+func (fh *finalizationHandler) start() {
+	defer func() {
+		close(fh.errorCh)
+		close(fh.handlerDone)
+	}()
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return
+		default:
+		}
+
+		err := fh.initiateRound()
+		if err != nil {
+			fh.errorCh <- fmt.Errorf("initiating round: %w", err)
+			return
+		}
+
+		err = fh.waitServices()
+		if err != nil {
+			fh.errorCh <- fmt.Errorf("waiting for services: %w", err)
+			return
+		}
+	}
+}
+
+func (fh *finalizationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	finalizationEngineErrCh := make(chan error)
+	go func() {
+		finalizationEngineErrCh <- fh.finalizationEngine.Stop()
+	}()
+
+	votingRoundErrCh := make(chan error)
+	go func() {
+		votingRoundErrCh <- fh.votingRound.Stop()
+	}()
+
+	finalizationEngErr := <-finalizationEngineErrCh
+	votingRoundErr := <-votingRoundErrCh
+
+	if finalizationEngErr != nil && votingRoundErr != nil {
+		return fmt.Errorf("%w: %s; %s", errServicesStopFailed, finalizationEngErr, votingRoundErr)
+	}
+
+	if finalizationEngErr != nil {
+		return fmt.Errorf("%w: %s", errServicesStopFailed, finalizationEngErr)
+	}
+
+	if votingRoundErr != nil {
+		return fmt.Errorf("%w: %s", errServicesStopFailed, votingRoundErr)
+	}
+
+	return nil
```

Comment:
```suggestion
	if finalizationEngErr != nil && votingRoundErr != nil {
		return fmt.Errorf("%w: %s; %s", errServicesStopFailed, finalizationEngErr, votingRoundErr)
	} else if finalizationEngErr != nil {
		return fmt.Errorf("%w: %s", errServicesStopFailed, finalizationEngErr)
	} else if votingRoundErr != nil {
		return fmt.Errorf("%w: %s", errServicesStopFailed, votingRoundErr)
	} else {
	        return nil
	}
```

Reply:
Both ways achieve the same result is there any specific reason for changing from the current way to use the `if {} else if {}` other than the style?